component/module/git/internal/vcs/auth/auth.go

```diff
@@ -0,0 +1,67 @@
+package auth
+
+import (
+	"fmt"
+
+	"github.com/go-git/go-git/v5/plumbing/transport"
+	"github.com/go-git/go-git/v5/plumbing/transport/http"
+	"github.com/go-git/go-git/v5/plumbing/transport/ssh"
+	"github.com/grafana/agent/pkg/river/rivertypes"
+)
+
+type GitAuthConfig struct {
+	BasicAuth *BasicAuth `river:"basic_auth,block,optional"`
+	SSHKey    *SSHKey    `river:"ssh_key,block,optional"`
+}
+
+// Convert converts HTTPClientConfig to the native Prometheus type. If h is
+// nil, the default client config is returned.
+func (h *GitAuthConfig) Convert() transport.AuthMethod {
+	if h == nil {
+		return nil
+	}
+
+	if h.BasicAuth != nil {
+		return h.BasicAuth.Convert()
+	}
+
+	if h.SSHKey != nil {
+		key, _ := h.SSHKey.Convert()
+		return key
+	}
+	return nil
+}
+
+type BasicAuth struct {
+	Username string            `river:"username,attr,optional"`
+	Password rivertypes.Secret `river:"password,attr,optional"`
+}
+
+// Convert converts our type to the native prometheus type
+func (b *BasicAuth) Convert() (t transport.AuthMethod) {
+	if b == nil {
+		return nil
+	}
+	return &http.BasicAuth{
+		Username: b.Username,
+		Password: string(b.Password),
+	}
+}
+
+type SSHKey struct {
+	Username   string            `river:"username,attr,optional"`
+	Keyfile    string            `river:"keyfile,attr,optional"`
+	Passphrase rivertypes.Secret `river:"passphrase,attr,optional"`
+}
+
+// Convert converts our type to the native prometheus type
+func (s *SSHKey) Convert() (transport.AuthMethod, error) {
+	if s == nil {
+		return nil, nil
+	}
+	publickeys, err := ssh.NewPublicKeysFromFile(s.Username, s.Keyfile, string(s.Passphrase))
+	if err != nil {
+		return nil, fmt.Errorf("Loading SSH keys failed: %s", err.Error())
+	}
+	return publickeys, nil
```

Comment:
Cool. I'd like to see it possible to specify the key from another component, which would be extra useful to get the SSH keys from Vault. It looks like the ssh package has a method to give it PEM bytes directly: https://pkg.go.dev/github.com/go-git/go-git/v5/plumbing/transport/ssh#NewPublicKeys

Could we make `key` an extra attribute which is mutually exclusive with setting `key_file`, and use NewPublicKeys if the `key` attribute is non-empty, and NewPublicKeysFromFile otherwise?

Reply:
I really like this idea. I added in a new 'key' attribute that uses `NewPublicKeys` instead and tested it with a `local.file` component. Worked really well in my local tests.